### PR TITLE
Improve model consistency

### PR DESCRIPTION
**Pull request by microbit-robert**
Model accuracy can be highly variable due to non-determinism. This change improves the consistency of the trained model.

### CODE REVIEW

File: src/script/ml.ts [hunk 1]
```diff
@@ -105,20 +107,31 @@ export async function trainModel(): Promise<tf.LayersModel | void> {
   const tensorFeatures = tf.tensor(features);
   const tensorLabels = tf.tensor(labels);
   const nn: LayersModel = createModel();
-  const totalNumEpochs = get(settings).numEpochs;
+  const totalNumEpochs = 160;
```
Comment (microbit-robert) [inline src/script/ml.ts:110]: Hardcode value here changing the initial ml settings won't affect anyone who has used the app before. Settings in local storage are preferentially used.

File: src/script/ml.ts [hunk 2]
```diff
@@ -105,20 +107,31 @@ export async function trainModel(): Promise<tf.LayersModel | void> {
...
-      },
+      callbacks: [
+        new tf.CustomCallback({ onTrainEnd }),
+        new tf.CustomCallback({
+          onEpochEnd(epoch, logs) {
+            // Epochs indexed at 0
+            updateTrainingProgress(epoch / (totalNumEpochs - 1));
+            if (logs) {
+              if (prevLoss && logs.loss && logs.accuracy) {
+                if (logs.loss >= prevLoss && logs.accuracy === 1) {
```
[the remark this answers is not in the document]
Comment (microbit-robert) [reply src/script/ml.ts]: Yep -> [055f312](https://github.com/microbit-foundation/ml-trainer/pull/270/commits/055f31206e45b31967fad79de3ed1c8a5884c15a).

I think in practice this is rarely triggered. We can overshoot the best fit, and once we realise we've done this, we can't go back to the best fit. What we could do with is the [built-in earlyStopping callback](https://js.tensorflow.org/api/latest/#callbacks.earlyStopping) to support `restoreBestWeights`, but it's not currently supported.

[the remark this answers is not in the document]
Comment (microbit-grace) [reply src/script/ml.ts]: Oh interesting, thanks for the info and link! Sad that it's not currently supported

I noticed that the built-in earlyStopping callback seems to monitor `val_acc` instead of `acc`. It makes me wonder whether we should be using `val_loss` and `val_acc` instead? 🤔

[the remark this answers is not in the document]
Comment (microbit-robert) [reply src/script/ml.ts]: It's `val_acc` by default, but you can change the metric to track. My understanding is that you should use `val_loss` and `val_acc` for regression problems, and `loss` and `acc` for classification problems. The [upstream version is now also tracking loss](https://github.com/microbit-foundation/cctd-ml-machine/blob/a90b7f1f1351d42f8958fc4d932b4201f283ef8a/src/script/mlmodels/LayersModelTrainer.ts#L76) to plot it.

[the remark this answers is not in the document]
Comment (microbit-grace) [reply src/script/ml.ts]: Ok, that makes sense, thanks!

**Review by microbit-grace — APPROVED** [review]
Not sure if Matt wants to have a look at it too, but LGTM

**Comment by microbit-robert** [comment]
@microbit-grace I think it's worth comparing to this branch which does use the built-in early stopping callback https://review-ml.microbit.org/model-consistency-alt.

It stops training very early when using the CCTD example dataset (still, shake, circle), which is used in the ml.test.ts file and causes the tests to fail (the resulting model isn't great). It seems to work well on other datasets though. So maybe this is a result of the training data not being that good. We've seen the model mistake still for circle.

**Comment by microbit-matt-hillsdon** [comment]
Folks will get this change along with the beta so no need for this at this point. Please shout if I've misunderstood.
